apis/python/src/tiledbsoma/_read_iters.py

```diff
@@ -5,59 +5,420 @@
+            d: pd.Index(self.joinids[d].to_numpy())
+            for d in (self.axes_to_reindex - set((self.major_axis,)))
+        }
+
+        # Ask subclass to create the type-specific reader/iterator
+        self._reader = self._create_reader()
+
+    @classmethod
+    def _validate_args(
+        cls,
+        shape: NTuple,
+        axis: Union[int, Sequence[int]],
+        size: Optional[Union[int, Sequence[int]]] = None,
+        reindex_disable: Optional[Union[int, Sequence[int]]] = None,
+    ) -> Tuple[List[int], List[int], List[int]]:
+        """
+        Class method to validate and normalize common user-provided arguments axis, size and reindex_disable.
+        * normalize args to fully specify the arg per dimension, for convenience in later usage
+        * error check and raise if a nonsense value
+        * set defaults where supported.
+        """
+        ndim = len(shape)
+
+        # convert to list
+        axis = list(axis if isinstance(axis, Sequence) else [axis])
+        reindex_disable = (
+            list(
+                reindex_disable
+                if isinstance(reindex_disable, Sequence)
+                else [reindex_disable]
+            )
+            if reindex_disable is not None
+            else []
+        )
+
+        # Currently, only support chunking on one dimension.
+        if len(axis) != 1:
+            raise NotImplementedError(
+                "Multi-dimension blockwise iterators not implemented"
+            )
+        # all dim indices must be in acceptable range
+        if not all(0 <= d < ndim for d in axis):
+            raise ValueError("blockwise `axis` value must be in range [0, ndim)")
+        if not all(0 <= d < ndim for d in reindex_disable):
+            raise ValueError(
+                "blockwise `reindex_disable` value must be in range [0, ndim)"
+            )
+
+        # if not specified, set default size for each axis. Default heuristic
+        # assumes 2D array has many more rows than cols (i.e., n_obs>>n_vars).
+        default_block_size = (2**16,) + (2**8,) * (ndim - 1)
+        if size is None:
+            size = [default_block_size[d] for d in axis]
+        elif isinstance(size, int):
+            size = [size] * len(axis)
+        elif isinstance(size, Sequence):
+            size = list(size) + [default_block_size[d] for d in axis[len(size) :]]
+        else:
+            raise TypeError(
+                "blockwise iterator `size` must be None, int or Sequence[int]"
+            )
+
+        return axis, size, reindex_disable
+
+    @abc.abstractmethod
+    def _create_reader(self) -> Iterator[_RT]:
+        """Sub-class responsibility"""
+        raise NotImplementedError()
+
+    def __next__(self) -> _RT:
+        return next(self._reader)
+
+    def concat(self) -> _RT:
+        """
+        Unimplemented as there is little utility beyond that offered by a ragged
+        read iterator concat, other than reindexing.
```

Comment:
Why not implement anyway? Seems like it would be a less surprising outcome for the user. And isn't reindexing a valid reason to enable this?

Reply:
My principle was to not add additional code/API that the user could replicate with a single line on their own (e.g., `pandas.Index(...).get_indexer(...)`.  So yeah, it is a valid case, but I don't see what value we are providing by making our own API contain more sugar....

Love to get other's thoughts on this

Reply:
OK, maybe something to consider for a SOMA API V2, where "Full" result read methods could be exposed at the `SparseRead` level, rather than at `ReadIter`.

Reply:
The other issue I forgot to mention -- `concat` is complicated when you consider reindexing on the primary (blockwise) axis.  Naively implemented, it would just create a bunch of duplicate coordinates (zero to N), so we probably would need to either disallow reindexing on the primary dimension when concat is called, _or_ redefine it as reindexing the full concatenated result.

For SciPy blockwise reads, there is already a trivial implementation that does the same, without the sharp edges, which is the `vstack()` method in the SciPy sparse package.  This allows nice patterns like:

```
with soma.open(a_random_sparse_nd_array, mode="r") as A:
    full_csr_matrix = sparse.vstack(sp for sp, _ in A.read(coords).blockwise(axis=0).scipy())
```